kernel/src/engine/arrow_utils.rs

```diff
@@ -1066,53 +1064,47 @@ pub(crate) fn parse_json(
-// StringArray -> StructArray JSON parsing. See https://github.com/apache/arrow-rs/issues/6522. If
-// that shortcoming gets fixed upstream, this method can simplify or hopefully even disappear.
 fn parse_json_impl(json_strings: &StringArray, schema: ArrowSchemaRef) -> DeltaResult<RecordBatch> {
     if json_strings.is_empty() {
         return Ok(RecordBatch::new_empty(schema));
     }
 
-    // Use batch size of 1 to force one record per string input
     let mut decoder = ReaderBuilder::new(schema.clone())
-        .with_batch_size(1)
+        .with_batch_size(json_strings.len())
         .build_decoder()?;
-    let parse_one = |json_string: Option<&str>| -> DeltaResult<RecordBatch> {
-        let mut reader = BufReader::new(json_string.unwrap_or("{}").as_bytes());
-        // loop to fill + empty the buffer until end of input. note that we can't just one-shot
-        // attempt to decode the entire thing since the buffer might only contain part of the JSON.
-        // see: https://github.com/delta-io/delta-kernel-rs/pull/1244
-        loop {
-            let buf = reader.fill_buf()?;
-            if buf.is_empty() {
-                break;
-            }
-            // from `decode` docs:
-            // > Read JSON objects from `buf`, returning the number of bytes read
-            // > This method returns once `batch_size` objects have been parsed since the last call
-            // > to [`Self::flush`], or `buf` is exhausted. Any remaining bytes should be included
-            // > in the next call to [`Self::decode`]
-            //
-            // if we attempt a `parse_one` of e.g. "{}{}", we will parse the first "{}" successfully
-            // then decode will always return immediately sinee we have read `batch_size = 1`,
-            // leading to an infinite loop. Since we always just want to parse one record here, we
-            // detect this by checking if we always consume the entire buffer, and error if not.
-            let consumed = decoder.decode(buf)?;
-            if consumed != buf.len() {
-                return Err(Error::generic("Malformed JSON: Multiple JSON objects"));
-            }
-            reader.consume(consumed);
+
+    for json in json_strings {
+        let line = json.unwrap_or("{}").trim();
+        if decoder.has_partial_record() {
+            return Err(Error::generic("Malformed JSON: Multiple JSON objects"));
         }
-        let Some(batch) = decoder.flush()? else {
-            return Err(Error::missing_data("Expected data"));
-        };
-        require!(batch.num_rows() == 1, Error::generic("Expected one row"));
-        Ok(batch)
-    };
-    let output: Vec<_> = json_strings.iter().map(parse_one).try_collect()?;
-    Ok(concat_batches(&schema, output.iter())?)
+        if line.is_empty() {
+            return Err(Error::missing_data(
+                "Malformed JSON: I don't know what to do with an empty string! {line:?}",
```

Comment:
```suggestion
                "Malformed JSON: empty string is not parseable",
```

Reply:
I also think this check is probably made redundant with the check suggested below with row_num (we check 2 cases: if `row_num + 1 > decoder.len` then there was an empty string or partial string, if `row_num + 1 < decoder.len` we have a duplicate object.
,